Opentitan EG 1.0.0 I2C implementation

ziuziakowska opened the pull request:
This PR implements the Opentitan Earlgrey 1.0.0 I2C controller, based on the current Darjeeling I2C implementation which to my understanding is similar to Earlgrey ES.

The main differences are a different register interface with some moved fields, most FIFO overflow interrupts are replaced with threshold interrupts, and FIFO thresholds can be any amount of bytes.

Unsupported or partially implemented features:

- Only 7-bit addressing is supported, with no address bit masking. (`TARGET_ID` only supports `ADDRESS0` with `MASK0` == `0x7f`)
- Loopback mode (`CTRL.LLPBK`) is unsupported.
- Timing-related registers or physical layer controls are unimplemented:
  - `TIMING0`-`TIMING4` registers. (besides the changes in `earlgrey-9.2.0` that at least check the values are within spec.)
  - `OVRD` to manually drive SDA/SCL. `VAL`, `ACQ_FIFO_NEXT_DATA`.
  - Clock stretching controls. (`HOST_TIMEOUT_CTRL`, `TARGET_TIMEOUT_CTRL`, `TARGET_ACK_CTRL`, `HOST_NACK_HANDLER_TIMEOUT`)

ziuziakowska committed:
[ot] hw/riscv: ot_earlgrey: add i2c plic irqs numbers

taken from `hw/top_earlgrey/sw/autogen/top_earlgrey.h`

Signed-off-by: Alice Ziuziakowska <a.ziuziakowska@lowrisc.org>

ziuziakowska committed:
[ot] hw/opentitan: ot_i2c: update interrupts and registers to eg 1.0.0

This updates the interrupts and registers for the I2C block from the
earlgrey-es interface to earlgrey 1.0.0.

`FIFO_STATUS` has been split off into `HOST_FIFO_STATUS` and
`TARGET_FIFO_STATUS`, and some fields from `FIFO_CTRL` have been moved
to the new `HOST_FIFO_CONFIG` and `TARGET_FIFO_CONFIG`.

The remaining newly added registers are not yet implemented.

Interrupts are now caused by exceeding a per-fifo threshold instead on
overflow.

Signed-off-by: Alice Ziuziakowska <a.ziuziakowska@lowrisc.org>

ziuziakowska committed:
[ot] hw/opentitan: ot_i2c: impl `HOST_FIFO_CONFIG`,`TARGET_FIFO_CONFIG`

Signed-off-by: Alice Ziuziakowska <a.ziuziakowska@lowrisc.org>

ziuziakowska committed:
[ot] hw/opentitan: ot_i2c: change host/target rx/tx to FIFO names

This commit changes the names of
`ot_i2c_{host,target}_get_{rx,tx}_threshold` to use the names of the
FIFOs it is checking. It also adds
`ot_i2c_{fmt,acq,rx,tx}_threshold_intr` functions to compare against the
threshold to determine if the corresponding level interrupt should be
asserted, as some FIFOs assert the interrupt below their threshold and
others above.

Signed-off-by: Alice Ziuziakowska <a.ziuziakowska@lowrisc.org>

ziuziakowska committed:
[ot] hw/opentitan: ot_i2c: make `FIFO_CTRL` register write-only

With the removal of the `FMTILVL` and `RXILVL` fields, this register is
now write-only.

Signed-off-by: Alice Ziuziakowska <a.ziuziakowska@lowrisc.org>

ziuziakowska committed:
[ot] hw/opentitan: ot_i2c: impl `CONTROLLER_EVENTS`,`TARGET_EVENTS`

These registers contain latched event bits that are RW1C. All bits must
be cleared to deassert the interrupt.

Signed-off-by: Alice Ziuziakowska <a.ziuziakowska@lowrisc.org>

ziuziakowska committed:
[ot] hw/opentitan: ot_i2c: add some additional comments and logging

Signed-off-by: Alice Ziuziakowska <a.ziuziakowska@lowrisc.org>

ziuziakowska committed:
[ot] hw/opentitan: ot_i2c: implement `FDATA.NAKOK`

This implements the `NAKOK` field when sending bytes in host mode.

If `NAKOK` is set, then the current byte in a write transaction
will not cause a controller halt interrupt if the byte is not ACK'd.
The behaviour if READB and NAKOK are both set is unspecified in the docs
so this is rejected.

Since this flag needs to be stored for each byte, the FMT FIFO is
upgraded from a `Fifo8` to a `OtFifo32` to store the extra bit.
This bit is not sent over I2C as it only affects controller interrupts.

Signed-off-by: Alice Ziuziakowska <a.ziuziakowska@lowrisc.org>

AlexJones0 commented:
I've not had a chance to look in detail yet, but one quick high-level comment: as I understand, although the *current* QEMU Darjeeling I2C implementation (i.e. what is already in `ot_i2c_dj.c`)  is similar to Earlgrey ES, the current OpenTitan Darjeeling I2C (i.e. what we want to emulate) should be similar to the current Earlgrey I2C.

So, I think in this PR it would be appropriate to just rename `ot_i2c_dj` to `ot_i2c` and use it for both tops, as I don't think there are any notable differences for emulation at the moment, so duplicating the device and having 2 I2C implementations will just be harder to maintain I think.

*Aside: there do seem to have been 1 or 2 RTL Changes since `earlgrey_1.0.0` was taken, but only to hardware internals like e.g. the DMA interface - these could be added to Darjeeling if necessary with versioning later on, like is done in e.g. the [pwrmgr](https://github.com/lowRISC/qemu/blob/ffc53c916f285f855533f810a350bb601c30f128/include/hw/opentitan/ot_pwrmgr.h#L37-L42).*

AlexJones0 commented:
Also another note: if you could add a commit that quickly updates the status of I2C in `docs/opentitan/earlgrey.md` that would be great. It doesn't have to be super descriptive, just mention any notable supported/missing features as is done for the other IP.

loiclefort commented:
> I think in this PR it would be appropriate to just rename ot_i2c_dj to ot_i2c and use it for both tops

Agreed. Earlgrey and Darjeeling should share the same I2C implementation.

ziuziakowska committed:
[ot] hw/opentitan: ot_i2c: remove `target_rx_nack`

The I2C_NACK event seems to be unreachable as we do not use
`i2c_send_async`, `i2c_send` returns a NACK by return value, `i2c_recv`
is infallible, and we ACK every byte we receive in target mode so we
cannot receive a byte that was NACK'd.

Signed-off-by: Alice Ziuziakowska <a.ziuziakowska@lowrisc.org>

ziuziakowska committed:
[ot] docs/opentitan: earlgrey: document I2C controller features

Signed-off-by: Alice Ziuziakowska <a.ziuziakowska@lowrisc.org>

engdoreis reviewed (commented):
Thanks @ziuziakowska for this PR.
I'm not very familiar with the code base, so I'll leave in deth reviews for Alex and Loïc.
The only thing that I beleive that would make sense to change is to squash the commits:
 - "[ot] hw/riscv: ot_earlgrey: copy ot_i2c_dj into ot_i2c, add ot_i2c to eg"
 - "[ot] hw: riscv,opentitan: replace ot_i2c_dj with ot_i2c"
 
I would also suggest running clang-format/tidy on every commit.

rivos-eblot commented on hw/opentitan/ot_i2c.c:
```diff
@@ -269,11 +310,15 @@ struct OtI2CDjState {
-    /* FMT: Scheduled operations for host mode. */
-    Fifo8 host_tx_fifo;
+    /*
+     * FMT: Shceduled operations for host mode.
```
typo

rivos-eblot commented on configs/devices/riscv32-softmmu/default.mak:
```diff
@@ -18,3 +18,6 @@ CONFIG_IBEXDEMO=y
 CONFIG_OT_DARJEELING=y
 CONFIG_OT_EARLGREY=y
 CONFIG_OPENTITAN=n
+CONFIG_I2C=y
```
I think it is a weird location to add these defintions,
It guess that if a dependency on I2C is required, it should be done from `CONFIG_OT_I2C`
I'm not sure about what `CONFIG_TEST_DEVICES` drags, but it might be a dependency on the OT machine itself.

In all cases, those are not dependencies on RISCV, but on OpenTitan. Someone who does not want to build OT machine(s) should not have to disable many lines from this file.

ziuziakowska committed:
[ot] hw/riscv: ot_earlgrey: move `ot_i2c_dj` to `ot_i2c`, add to eg

Signed-off-by: Alice Ziuziakowska <a.ziuziakowska@lowrisc.org>

ziuziakowska committed:
[ot] hw/opentitan: ot_i2c: use `ot_id` in I2C bus name

This commit makes each I2C controller create its own I2C bus using its
`ot_id` in the bus name ('ot-i2c0`, `ot-i2c1`, `ot-i2c2`).
Previously, each instantiation would overwrite the previous and leave
only one `ot-i2c` bus.

Signed-off-by: Alice Ziuziakowska <a.ziuziakowska@lowrisc.org>

rivos-eblot commented on hw/opentitan/ot_i2c.c:
```diff
@@ -881,7 +978,13 @@ static void ot_i2c_dj_write(void *opaque, hwaddr addr, uint64_t val64,
         address = FIELD_EX32(val32, TARGET_ID, ADDRESS0);
         mask = FIELD_EX32(val32, TARGET_ID, MASK0);
-        if (address != 0 && mask == 0x7f) {
+        if (mask != 0x7f) {
```
0x7fu

rivos-eblot commented on hw/opentitan/ot_i2c.c:
```diff
@@ -881,7 +978,13 @@ static void ot_i2c_dj_write(void *opaque, hwaddr addr, uint64_t val64,
-        if (address != 0 && mask == 0x7f) {
+        if (mask != 0x7f) {
+            qemu_log_mask(LOG_GUEST_ERROR,
+                          "%s: %s: Mask != 0x7f is not supported.\n", __func__,
```
I think a more detailed error message would be useful.

rivos-eblot commented on hw/opentitan/ot_i2c.c:
```diff
@@ -967,9 +1080,27 @@ static void ot_i2c_dj_write(void *opaque, hwaddr addr, uint64_t val64,
         s->check_timings = true;
         break;
+    case R_CONTROLLER_EVENTS:
+        val32 &= 0xf; // RW1C mask
```
Prefer unsigned constants (several occurrences)

rivos-eblot commented on hw/opentitan/ot_i2c.c:
```diff
@@ -967,9 +1080,27 @@ static void ot_i2c_dj_write(void *opaque, hwaddr addr, uint64_t val64,
+        ot_i2c_irq_set_state(s, CONTROLLER_HALT, s->regs[reg] != 0);
+        break;
+    case R_TARGET_EVENTS:
+        val32 &= 0x7; // RW1C mask
```
Do not use hardcoded values, prefer #define

rivos-eblot reviewed (changes requested):
Note: I'm not familiar enough with the I2C implementation to provide feedback about HW emulation, so comments are mostly about generic QEMU OT idioms.

rivos-eblot commented on hw/riscv/Kconfig:
```diff
@@ -99,6 +100,9 @@ config OT_EARLGREY
     select OT_TIMER
     select OT_UART
     select OT_UNIMP
+    select I2C
```
Still not the best location: generic I2C should be a dependency of OT_I2C + why would it be enabled for EG and not DJ?

rivos-eblot commented on hw/riscv/Kconfig:
```diff
@@ -99,6 +100,9 @@ config OT_EARLGREY
     select OT_UNIMP
+    select I2C
+    select I2C_DEVICES
+    select TEST_DEVICES
```
Rule should be the same for EG & DJ; I think `imply` better fits that `select`: this is not a hard dependency. It may be worth to check what we actually need. Is it only `I2C_ECHO`?

rivos-eblot commented on hw/opentitan/ot_i2c.c:
```diff
@@ -134,67 +137,99 @@ REG32(FDATA, 0x1cu)
+     INTR_ACQ_THRESHOLD_MASK | INTR_CONTROLLER_HALT_MASK | \
+     INTR_TX_STRETCH_MASK | INTR_TX_THRESHOLD_MASK | INTR_ACQ_STRETCH_MASK)
+
+#define CONTROLLER_EVENTS_RW1C_MASK 0xfu
```
If my understanding is correct about those constants, they would better be defined as an explicit combination of the CONTROLLER_EVENTS / TARGET_EVENTS bits.

rivos-eblot commented on hw/opentitan/ot_i2c.c:
```diff
@@ -881,7 +981,14 @@ static void ot_i2c_dj_write(void *opaque, hwaddr addr, uint64_t val64,
         address = FIELD_EX32(val32, TARGET_ID, ADDRESS0);
         mask = FIELD_EX32(val32, TARGET_ID, MASK0);
-        if (address != 0 && mask == 0x7f) {
+        if (mask != 0x7fu) {
```
Replace mask hardcoded value with a constant (likely defined from some existing definitions?)

AlexJones0 commented on hw/opentitan/ot_i2c.c:
```diff
@@ -134,67 +137,107 @@ REG32(FDATA, 0x1cu)
+    FIELD(TIMING3, THD_DAT, 16u, 12u)
+REG32(TIMING4, 0x4cu)
+    FIELD(TIMING4, TSU_STO, 0u, 12u)
+    FIELD(TIMING4, T_BUF, 16u, 12u)
```
I think that these field sizes don't match the docs?

```suggestion
REG32(TIMING0, 0x3cu)
    FIELD(TIMING0, THIGH, 0u, 13u)
    FIELD(TIMING0, TLOW, 16u, 13u)
REG32(TIMING1, 0x40u)
    FIELD(TIMING1, T_R, 0u, 10u)
    FIELD(TIMING1, T_F, 16u, 9u)
REG32(TIMING2, 0x44u)
    FIELD(TIMING2, TSU_STA, 0u, 13u)
    FIELD(TIMING2, THD_STA, 16u, 13u)
REG32(TIMING3, 0x48u)
    FIELD(TIMING3, TSU_DAT, 0u, 9u)
    FIELD(TIMING3, THD_DAT, 16u, 13u)
REG32(TIMING4, 0x4cu)
    FIELD(TIMING4, TSU_STO, 0u, 13u)
    FIELD(TIMING4, T_BUF, 16u, 13u)
```

AlexJones0 commented on hw/opentitan/ot_i2c.c:
```diff
@@ -134,67 +137,107 @@ REG32(FDATA, 0x1cu)
+REG32(TARGET_NACK_COUNT, 0x68u)
+    FIELD(TARGET_NACK_COUNT, TARGET_NACK_COUNT, 0u, 8u)
+REG32(TARGET_ACK_CTRL, 0x6cu)
+    FIELD(TARGET_ACK_CTRL, NBYTES, 0u, 8u)
```
```suggestion
    FIELD(TARGET_ACK_CTRL, NBYTES, 0u, 9u)
```

AlexJones0 commented on hw/opentitan/ot_i2c.c:
```diff
@@ -134,67 +137,107 @@ REG32(FDATA, 0x1cu)
+    FIELD(ACQ_FIFO_NEXT_DATA, ACQ_FIFO_NEXT_DATA, 0u, 8u)
+REG32(HOST_NACK_HANDLE_TIMEOUT, 0x74u)
+    FIELD(HOST_NACK_HANDLE_TIMEOUT, VAL, 0u, 31u)
+    FIELD(HOST_NACK_ENABLE_TIMEOUT, EN, 31u, 1u)
```
```suggestion
REG32(HOST_NACK_HANDLER_TIMEOUT, 0x74u)
    FIELD(HOST_NACK_HANDLER_TIMEOUT, VAL, 0u, 31u)
    FIELD(HOST_NACK_HANDLER_TIMEOUT, EN, 31u, 1u)
```

AlexJones0 commented on hw/opentitan/ot_i2c.c:
```diff
@@ -676,24 +760,26 @@ static uint64_t ot_i2c_dj_read(void *opaque, hwaddr addr, unsigned size)
+         * threshold. */
+        ot_i2c_irq_set_state(s, ACQ_THRESHOLD, ot_i2c_acq_threshold_intr(s));
+        break;
+    case R_HOST_FIFO_STATUS:
```
Nit: really the masking here and in `R_TARGET_FIFO_STATUS` reads is not necessary due to the FIFO capacity, and if we did want to check this it might be more appropriate as a `g_assert`? (because it would mean something has gone horribly wrong about our understanding of the FIFOs).

AlexJones0 commented on hw/opentitan/ot_i2c.c:
```diff
@@ -212,55 +255,68 @@ static const char *REG_NAMES[REGS_COUNT] = {
 #undef IRQ_NAME_ENTRY
 
-#define OT_I2C_DJ_FIFO_SIZE 64u
+#define OT_I2C_FIFO_SIZE 64u
```
To give you more work: looking at OpenTitan's I2C, the acquisition FIFO [actually has a different size of 268 bytes compared to the 64 bytes of the other FIFOs](https://github.com/lowRISC/opentitan/blob/867c06ac88df09dd9fca6eca6fe1e9966da61e11/hw/ip/i2c/data/i2c.hjson#L155-L169). Annoyingly, if this is documented anywhere I can't find it - maybe we should open an issue/PR for improving the I2C documentation?

If this change is going to be painful it might be a good idea to leave it to a future PR - after all, the I2C is already outdated, so there is no regression in this PR.

AlexJones0 commented on hw/opentitan/ot_i2c.c:
```diff
@@ -1087,58 +1226,58 @@ static uint8_t ot_i2c_dj_target_recv(I2CSlave *target)
     DEFINE_PROP_END_OF_LIST(),
 };
 
 static void ot_i2c_dj_reset_enter(Object *obj, ResetType type)
```
Nit: you missed one rename on this `ot_i2c_dj_reset_enter` (and when it is registered as a callback).

AlexJones0 commented on hw/opentitan/ot_i2c.c:
```diff
@@ -1265,8 +1265,8 @@ static void ot_i2c_realize(DeviceState *dev, Error **errp)
-    /* TODO: check if the following can be moved to ot_i2c_dj_init */
-    s->bus = i2c_init_bus(dev, TYPE_OT_I2C);
+    /* TODO: check if the following can be moved to ot_i2c_init */
+    s->bus = i2c_init_bus(dev, s->ot_id);
```
(very minor) nit: I would prefer prefixing the type name to the bus, i.e.:

```c
     char *bus_name = g_strdup_printf(TYPE_OT_I2C "-%s", s->ot_id);
     s->bus = i2c_init_bus(dev, bus_name);
     g_free(bus_name);
```

But I'll leave this up to your preference.

rivos-eblot replied:
(note: if renamed, do not forget to update the docs as well)

AlexJones0 commented on hw/opentitan/ot_i2c.c:
```diff
@@ -967,9 +1092,27 @@ static void ot_i2c_dj_write(void *opaque, hwaddr addr, uint64_t val64,
         break;
+    case R_CONTROLLER_EVENTS:
+        val32 &= CONTROLLER_EVENTS_RW1C_MASK;
+        s->regs[reg] &= ~val32;
```
Minor nit, just convention in the OpenTitan QEMU code: can you add a comment like:

```suggestion
        s->regs[reg] &= ~val32; /* RW1C */
```

on these RW1C fields so it is clear to someone casually browsing why the write is implemented this way?

AlexJones0 commented on hw/opentitan/ot_i2c.c:
```diff
@@ -972,9 +979,16 @@ static void ot_i2c_write(void *opaque, hwaddr addr, uint64_t val64,
+                "%s: %s: Address Mask with any bits unset is not supported.\n",
+                __func__, s->ot_id);
+            break;
+        }
```
Is this meant to be `LOG_UNIMP`?

AlexJones0 commented on hw/opentitan/ot_i2c.c:
```diff
@@ -722,17 +732,10 @@ static uint64_t ot_i2c_read(void *opaque, hwaddr addr, unsigned size)
         /* Report host TX FIFO status. */
-        if (fifo8_is_empty(&s->host_tx_fifo)) {
-            /* Used by drivers to see if partial transactions are done. */
-            if (!i2c_bus_busy(s->bus)) {
```
Quick question (maybe I'm just not understanding) - why do we no longer check that the I2C bus isn't busy?

ziuziakowska replied:
In QEMU, the I2C bus is "busy" if there is a transfer occurring. With this check, an infinite loop occurs when using `i2c_testutils_read` as it calls `i2c_testutils_wait_transaction_finish` which waits for the FMT FIFO to be empty. But this FIFO will never report as so because the bus is "busy" with this transaction.

AlexJones0 commented on hw/opentitan/ot_i2c.c:
```diff
@@ -769,6 +867,12 @@ static void ot_i2c_dj_write_fdata(OtI2CDjState *s, uint32_t fdata)
+            qemu_log_mask(LOG_GUEST_ERROR,
+                          "%s: %s: Invalid FDATA flags READB+NAKOK\n", __func__,
+                          s->ot_id);
+            return;
```
This is actually a bit more complicated since in reality we aren't emulating the full I2C FSM in enough detail - looking at the RTL, I think these signals are either ignored if you are not in the correct controller FSM state, or produce some SDA/SCL signals that are invalid (outside-the-spec).

For emulation, to closely match the HW, I think it makes the most sense if the I2C controller just ignores the NAKOK bit and processes the command as if it wasn't set (i.e. it just does a READB). The guest error is valuable, but we probably shouldn't return here.

```suggestion
```

The existing check of `READB+RCONT+STOP` is a bit more complicated - this seems like it would definitely produce invalid I2C signals, so I think a guest error and returning here makes sense given what we can achieve inside QEMU.

AlexJones0 commented on hw/opentitan/ot_i2c.c:
```diff
@@ -1186,21 +1182,13 @@ static int ot_i2c_target_event(I2CSlave *target, enum i2c_event event)
         i2c_ack(s->bus);
         break;
     case I2C_NACK:
-        s->target_rx_nack = true;
```
This makes sense - can you add a `g_assert_not_reached()` or a log or something similar so that we're actually sure we're never hitting this case, and can diagnose the issue quickly if we somehow do with a future change?

AlexJones0 reviewed (commented):
Thanks for doing this @ziuziakowska, sorry it took me a while to get to.

I feel like I'm lacking the background to comment on the correct build configuration options so I'll leave that to @rivos-eblot, but I've left some comments on both the HW emulation and general QEMU idioms.

rivos-eblot commented on hw/opentitan/ot_i2c.c:
```diff
@@ -418,16 +495,23 @@ static uint8_t ot_i2c_dj_host_read_rx_fifo(OtI2CDjState *s)
-            ot_i2c_dj_irq_set_state(s, NAK, true);
+    while (!ot_fifo32_is_empty(&s->host_tx_fifo)) {
+        uint32_t val = ot_fifo32_pop(&s->host_tx_fifo);
+        uint8_t fbyte = FIELD_EX32(val, FDATA, FBYTE);
```
would be nice to add explicit casts for these two lines.

rivos-eblot commented on hw/opentitan/ot_i2c.c:
```diff
@@ -1001,61 +1144,54 @@ static void ot_i2c_dj_target_set_acqdata(OtI2CDjState *s, uint32_t data,
     case I2C_START_SEND_ASYNC:
         /* Set the first byte to the target address + RW bit as 0. */
-        ot_i2c_dj_target_set_acqdata(s, target->address << 1, SIGNAL_START);
+        ot_i2c_target_set_acqdata(s, target->address << 1, SIGNAL_START);
```
1u

rivos-eblot commented:
> I feel like I'm lacking the background to comment on the correct build configuration options so I'll leave that to @rivos-eblot, but I've left some comments on both the HW emulation and general QEMU idioms.

Build config seems ok to me, thanks for the changes.

AlexJones0 commented on hw/opentitan/ot_i2c.c:
```diff
@@ -134,67 +137,106 @@ REG32(FDATA, 0x1cu)
+    FIELD(ACQ_FIFO_NEXT_DATA, ACQ_FIFO_NEXT_DATA, 0u, 8u)
+REG32(HOST_NACK_HANDLER_TIMEOUT, 0x74u)
+    FIELD(HOST_NACK_HANDLER_TIMEOUT, VAL, 0u, 31u)
+    FIELD(HOST_NACK_ENABLER_TIMEOUT, EN, 31u, 1u)
```
```suggestion
    FIELD(HOST_NACK_HANDLER_TIMEOUT, EN, 31u, 1u)
```

rivos-eblot replied:
There's also a typo here:
`FIELD(TARGET_TIMOUT_CTRL, VAL, 0u, 31u)`

AlexJones0 reviewed (approved):
Thanks for the changes - I'm happy to merge this after my comment is addressed.

@rivos-eblot is this okay to merge for you? 
(edit: nevermind, I can't because of the requested changes anyway :sweat_smile:)

rivos-eblot reviewed (approved):
LGTM

rivos-eblot commented:
> @rivos-eblot is this okay to merge for you? (edit: nevermind, I can't because of the requested changes anyway 😅)
Yes thanks.

ziuziakowska committed:
[ot] hw/opentitan: ot_i2c: update ACQ FIFO size

Default depth value for this FIFO is 268.

See `hw/ip/i2c/data/i2c.hjson` in Opentitan.

Signed-off-by: Alice Ziuziakowska <a.ziuziakowska@lowrisc.org>
